action.php

```diff
@@ -216,68 +221,100 @@ function admin_change() {
+        $searchForm->addTagClose('div', ++$currElemPos);
+
+        // options
+        $options = [
+            'or' => $this->getLang('search_any_tag'),
+            'and' => $this->getLang('search_all_tags'),
+            ];
+        $searchForm->addTagOpen('ul', ++$currElemPos)->attr('aria-expanded', 'false');
+        foreach ($options as $key => $label) {
+            $listItem = $searchForm->addTagOpen('li', ++$currElemPos);
+            if ($active && $key === $active) {
+                $listItem->addClass('active');
+            }
+            $link = $this->getSettingsLink($label, 'taggings', $key);
+            $searchForm->addHTML($link, ++$currElemPos);
+            $searchForm->addTagClose('li', ++$currElemPos);
+        }
+        $searchForm->addTagClose('ul', ++$currElemPos);
+
+        $searchForm->addTagClose('div', ++$currElemPos);
+    }
     /**
      * Show tagged pages on searches
      *
      * @param $event
      * @param $param
      */
-    function echo_searchresults(Doku_Event &$event, $param) {
+    public function echo_searchresults(Doku_Event $event, $param) {
         global $ACT;
         global $QUERY;
+        global $INPUT;
 
         if ($ACT !== 'search') {
             return;
         }
 
+        /** @var helper_plugin_tagging $hlp */
+        $hlp = plugin_load('helper', 'tagging');
+
         // parse the search query and use the first found word as term
         $terms = ft_queryParser(idx_get_indexer(), $QUERY);
 
-        $tag = '';
-        if (isset($terms['phrases'][0])) {
-            $tag = $terms['phrases'][0];
-        } else {
-            if (isset($terms['and'][0])) {
-                $tag = $terms['and'][0];
-            }
-        }
+        $tag = $hlp->getTags($terms);
         if (!$tag) {
             return;
         }
 
-        // create filter from term and namespace
-        $filter = array('tag' => $tag);
-        if (isset($terms['ns'][0])) {
-            $filter['pid'] = $terms['ns'][0];
-            if (substr($filter['pid'], -1) !== ':') {
-                $filter['pid'] .= ':';
-            }
-            $filter['pid'] .= '*';
-        }
-        if (isset($terms['notns'][0])) {
-            $i = 0;
-            foreach ($terms['notns'] as $notns) {
-
-                if (substr($notns, -1) !== ':') {
-                    $notns .= ':';
-                }
-                $notns .= '*';
-                $filter['notpid' . $i] = $notns;
-                ++$i;
-            }
-
-        }
-
-        /** @var helper_plugin_tagging $hlp */
-        $hlp = plugin_load('helper', 'tagging');
-        $pages = $hlp->findItems($filter, 'pid');
+        $pages = $hlp->searchPages();
         if (!count($pages)) {
             return;
         }
 
         // create output HTML
+        // format tag search terms
+        $operator = ($INPUT->has('taggings') && $INPUT->str('taggings') === 'and') ? $this->getLang('search_all_label') : $this->getLang('search_any_label');
```

Comment:
no need to check for existence, $INPUT methods do that for you and return defaults if things aren't set.

PS: try to keep line lengths <120 chars

Reply:
fixed